Mathlib/Data/FinEnum.lean

```diff
@@ -56,6 +56,54 @@ def ofList [DecidableEq α] (xs : List α) (h : ∀ x : α, x ∈ xs) : FinEnum
 def toList (α) [FinEnum α] : List α :=
   (List.finRange (card α)).map (equiv).symm
 
+/-- Any two enumerations of the same type have the same length. -/
+theorem card_eq_card {α : Type u} (e₁ e₂ : FinEnum α) : e₁.card = e₂.card :=
+  Fin.equiv_iff_eq.mp ⟨e₁.equiv.symm.trans e₂.equiv⟩
+
+/-- Any enumeration of an empty type has length 0. -/
+theorem card_eq_zero_of_IsEmpty {α : Type u} [FinEnum α] [IsEmpty α] : card α = 0 :=
+  match h : card α with
+  | 0 => rfl
+  | _ + 1 => ‹IsEmpty α›.elim ((h ▸ equiv).symm 0)
+
+/-- Any enumeration of a non-empty type has a length with an actual predecessor. -/
+def cardPredOfNonempty (α : Type u) [FinEnum α] [Nonempty α] : { n : ℕ // card α = n + 1 } :=
+  match h : card α with
+  | 0 => False.elim <| ‹Nonempty α›.elim ((h ▸ equiv) · |>.elim0)
+  | n + 1 => ⟨n, rfl⟩
```

Comment:
This is a bit weird given that the predecessor is explicitly given as `card α - 1`. The idiomatic way of finding the predecessor of `a` is not to inhabit `{n : ℕ // a = n + 1}` but rather to prove `0 < a` and use `a - 1`. Hence I would suggest
```suggestion
lemma card_pos {α : Type*} [FinEnum α] [Nonempty α] : 0 < card α := sorry
lemma card_ne_zero {α : Type*} [FinEnum α] [Nonempty α] : card α ≠ 0 := card_pos.ne'
```